### PR TITLE
Features/425 resplit rework

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

Changes the resplit function to work with a tiling function and not send tiles around. Will likely not be as efficient as an Alltoallv but it maintains the order

Issue/s resolved: #425 #476 

## Changes proposed:
- create a tiling class which has tile divisions along what would be the split axis in every dimension
- resplit between two axes now uses a home-brew approach where tiles are communicated (no alltoallv)
- both In-place and out-of-place resplit now requires that the free space on node is equal at least twice the local data size.

## Type of change

Remove irrelevant options:
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Documentation update

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/tiling.py [hunk 1]
```diff
@@ -1,8 +1,254 @@
...
+        # tile_ends_g has the end points, need to get the start and stop
+        arb_slices = [None] * arr.numdims
+        if isinstance(key, int):
+            key = [key]
+        if len(key) < arr.numdims or key[-1] is None:
+            lkey = list(key)
+            lkey.extend([slice(0, None)] * (arr.numdims - len(key)))
+            key = lkey
+        for d in range(arr.numdims):
+            # todo: implement advanced indexing (lists of positions to iterate through
```
Comment (Markus-Goetz) [inline heat/core/tiling.py:194]: Please transform into issue if generally approved

File: heat/core/tiling.py [hunk 2]
```diff
@@ -1,8 +1,254 @@
...
+        >>> a.create_split_tiles()
+        >>> a.tiles[1]
+        [0] tensor([[14., 15., 16.],
+        [0]         [21., 22., 23.]])
+        [1] tensor([[17., 18.],
+        [1]         [24., 25.]])
+        [2] tensor([[19., 20.],
+        [2]         [26., 27.]])
+        """
+        # todo: strides can be implemented with using a list of slices for each dimension
```
Comment (Markus-Goetz) [inline heat/core/tiling.py:177]: Please turn into issue if generally approved

File: heat/core/dndarray.py [hunk 3]
```diff
@@ -2473,50 +2493,84 @@ def resplit_(self, axis=None):
...
+            new_locs = torch.where(new_arr.tiles.tile_locations == rpr)
+            new_locs = torch.stack([new_locs[i] for i in range(self.numdims)], dim=1)
+
+            for i in range(new_locs.shape[0]):
+                key = tuple(new_locs[i].tolist())
+                spr = self.tiles.tile_locations[key].item()
+                to_send = self.tiles[key]
+                if spr == rank and spr != rpr:
+                    waits.append(self.comm.Isend(to_send.clone(), dest=rpr, tag=rank))
+                elif spr == rpr == rank:
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py:2554]: I am not sure that this is what you want here. This can only evaluate to True on rank 1 and above

File: heat/core/tiling.py [hunk 4]
```diff
@@ -1,8 +1,254 @@
...
+        # create a tensor for the process rank of all the tiles
+        tile_locations = self.set_tile_locations(split=arr.split, tile_dims=tile_dims, arr=arr)
+
+        self.__DNDarray = arr
+        self.__tile_locations = tile_locations.int()
+        self.__tile_ends_g = tile_ends_g
+        self.__tile_dims = tile_dims
+
+    @staticmethod
+    def set_tile_locations(split, tile_dims, arr):
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/tiling.py:90]: it is external for the resplit function to generate where the tiles will go

File: heat/core/tiling.py [hunk 5]
```diff
@@ -1,8 +1,254 @@
...
+        #  2. get the split axis numbers for the other axes
+        #  3. build tile map
+        lshape_map = arr.create_lshape_map()
+        tile_dims = torch.zeros((arr.numdims, arr.comm.size), device=arr.device.torch_device)
+        if arr.split is not None:
+            tile_dims[arr.split] = lshape_map[..., arr.split]
+        w_size = arr.comm.size
+        for ax in range(arr.numdims):
+            if arr.split is None or not ax == arr.split:
+                size = arr.gshape[ax]
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/tiling.py:72]: to use that function we would need to loop over the number of ranks as well. although that function could be changed to use this method, then it could be used

File: heat/core/dndarray.py [hunk 6]
```diff
@@ -2425,17 +2428,34 @@ def __redistribute_shuffle(self, snd_pr, send_amt, rcv_pr, snd_dtype):
             if snd_pr > rcv_pr:  # data passed from a higher rank (append to bottom)
                 self.__array = torch.cat((self.__array, data), dim=self.split)
 
-    def resplit_(self, axis=None):
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/dndarray.py:2428]: because this function is determined to be inplace by the underscore. (i believe that this comment may be outdated though)

[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/dndarray.py:2428]: sorry i missed this, fixing now

**Comment by ClaudiaComito** [comment]
OK, I've fleetingly looked at the code, and run a few tests resplitting a 50,000x50,000 float tensor from 1 to 0. On 15 nodes this implementation isn't that much slower than the Alltoallv implementation. I'm for approving now, improving - esp. because of the space requirements - later. We need a working resplit. Thanks @coquelin77 !
